### PR TITLE
Reuse provider build for testing

### LINKED ISSUE #967: Test the provider binaries we're planning to release
Currently, we build one provider binary during prerequisites which is captured as an asset and used for running our integration tests against. However, we then discard this binary and use goreleaser to build brand new binaries at the point of release which are not tested.

We should build the provider binaries for all platforms at the beginning of the workflow, capture them as asstes, use them for integration tests, then publish the _tested_ provider binaries.

This will either require the removal of goreleaser (see Azure Native for how to release without goreleaser) or pay for their premium product.

### LINKED ISSUE #1033: Test provider binaries for publishing
Use the provider binary from the `build_provider` workflow when running tests rather than building a throw-away binary in pre-requisites.

It might be helpful to call `go generate` as part of the provider build to avoid requiring a dependency on the `prerequisites` job. We only need the version calculated before we can build the provider binary as the schema is already committed. We can verify that the committed schema is up to date in a standalone job.

### PR DESCRIPTION
**Pull request by t0yv0**
Fixes #967 
Fixes #1033

Unlike the ticket spec in #967 , only linux-amd64 binary is tested by the full set of integration tests. We might consider opening a follow up to smoke-test every single one of the platform builds, or else keep #967 open. 

As to the ticket spec in #1033 the dependency from build_provider to prerequisites is real because prerequisites builds the schema and the schema is needed in the binaries. This needs to stay.

Test run: https://github.com/pulumi/pulumi-azuread/pull/1589

Additional exploration: in prerequisites we run `make test_provider` which seems to try running unit tests in the `./provider` directory. Unfortunately some tests in `./provider` directory are also integration tests, e.g. in pulumi-aws and pulumi-azuread, and the convention of passing `go test -short` to only run the unit tests does not seem to be observed universally (pulumi-azuread breaks for example). For this reason removing `make provider` from prerequisites is not yet possible as it breaks our builds. This is likely out of scope of this PR.

### CODE REVIEW

File: provider-ci/internal/pkg/templates/bridged-provider/.github/actions/download-bin/action.yml [hunk 1]
```diff
@@ -1,16 +1,23 @@
-name: Download binary assets
-description: Downloads the provider and tfgen binaries to `bin/`.
+name: Download the provider binary
```
Comment (t0yv0) [inline provider-ci/internal/pkg/templates/bridged-provider/.github/actions/download-bin/action.yml:1]: This action is repurposed to only download the provider binary (not tfgen). It does that by unpacking the actual production build for linux-amd64 and not the custom "for-test" build.

File: provider-ci/internal/pkg/templates/bridged-provider/.github/actions/download-tfgen/action.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+name: Download the tfgen binary
```
Comment (t0yv0) [inline provider-ci/internal/pkg/templates/bridged-provider/.github/actions/download-tfgen/action.yml:1]: This action is split out from download-bin to download the tfgen binary only.

File: provider-ci/internal/pkg/templates/bridged-provider/.github/actions/upload-bin/action.yml [hunk 3]
```diff
@@ -1,15 +0,0 @@
-name: Upload bin assets
```
Comment (t0yv0) [inline provider-ci/internal/pkg/templates/bridged-provider/.github/actions/upload-bin/action.yml:1]: This action is removed and inlined in favor of a simple UploadArtifact invocation for the tfgen binary.

File: provider-ci/internal/pkg/templates/bridged-provider/.github/workflows/build_sdk.yml [hunk 4]
```diff
@@ -49,8 +49,8 @@ jobs:
           tools: pulumictl, pulumicli, ${{ matrix.language }}
       - name: Prepare local workspace
         run: make prepare_local_workspace
-      - name: Download bin
-        uses: ./.github/actions/download-bin
+      - name: Download tfgen
```
Comment (t0yv0) [inline provider-ci/internal/pkg/templates/bridged-provider/.github/workflows/build_sdk.yml:52]: This only needs the tfgen binary.

**Review by blampe — APPROVED** [review]

**Pull request merged**